File: Coding/python/deck_of_cards.py
"""Deck of cards (object-oriented design exercise).

Port of java/DeckOfCards.java (original package code.ds), which models a generic
Deck<T extends Card>, an abstract Card, a Suit enum and a Hand<T>.

The Java file has no concrete Card subclass and no main; a SimpleCard subclass
and a small demo are added here so the module is runnable.

NOTE: Deck.shuffle is ported faithfully and is effectively a no-op: Java computes
`(int) Math.random() % (i + 1)`, and `(int)` of a value in [0, 1) is always 0, so
every swap index j is 0.
"""

from __future__ import annotations

import enum
import random
from typing import List, Optional


class Suit(enum.Enum):
    CLUB = 0
    DIAMOND = 1
    HEART = 2
    SPADE = 3

    @staticmethod
    def get_suit_from_value(value: int) -> Optional["Suit"]:
        for s in Suit:
            if s.value == value:
                return s
        return None


class Card:
    """Abstract card; subclasses must implement value()."""

    def __init__(self, c: int, s: Suit):
        self._available = True
        # number/face on the card: 2-10, 11 = Jack, 12 = Queen, 13 = King, 1 = Ace.
        self.face_value = c
        self.suit_ = s

    def value(self) -> int:
        raise NotImplementedError("abstract method: value()")

    def suit(self) -> Suit:
        return self.suit_

    def is_available(self) -> bool:
        """Whether the card is available to be given out to someone."""
        return self._available

    def mark_unavailable(self) -> None:
        self._available = False

    def mark_available(self) -> None:
        self._available = True

    def print_card(self) -> None:
        face_values = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10",
                       "J", "Q", "K"]
        out = face_values[self.face_value - 1]
        if self.suit_ == Suit.CLUB:
            out += "c"
        elif self.suit_ == Suit.HEART:
            out += "h"
        elif self.suit_ == Suit.DIAMOND:
            out += "d"
        elif self.suit_ == Suit.SPADE:
            out += "s"
        print(out + " ", end="")


class SimpleCard(Card):
    """Concrete Card whose value is just its face value (added for the demo)."""

    def value(self) -> int:
        return self.face_value


class Deck:
    """A generic deck of cards."""

    def __init__(self):
        self.cards: List[Card] = []
        self.dealt_index = 0  # marks first undealt card

    def set_deck_of_cards(self, deck_of_cards: List[Card]) -> None:
        self.cards = deck_of_cards

    def shuffle(self) -> None:
        # Faithful port: j is always 0 (see the module docstring), so this just
        # repeatedly swaps cards[i] with cards[0].
        for i in range(len(self.cards) - 1, 0, -1):
            j = int(random.random()) % (i + 1)
            card1 = self.cards[i]
            card2 = self.cards[j]
            self.cards[i] = card2
            self.cards[j] = card1

    def remaining_cards(self) -> int:
        return len(self.cards) - self.dealt_index

    def deal_hand(self, number: int) -> Optional[List[Optional[Card]]]:
        if self.remaining_cards() < number:
            return None
        hand: List[Optional[Card]] = [None] * number
        count = 0
        while count < number:
            card = self.deal_card()
            if card is not None:
                hand[count] = card
                count += 1
        return hand

    def deal_card(self) -> Optional[Card]:
        if self.remaining_cards() == 0:
            return None
        card = self.cards[self.dealt_index]
        card.mark_unavailable()
        self.dealt_index += 1
        return card

    def print_deck(self) -> None:
        for card in self.cards:
            card.print_card()


class Hand:
    """A hand of cards with a simple additive score."""

    def __init__(self):
        self.cards: List[Card] = []

    def score(self) -> int:
        return sum(card.value() for card in self.cards)

    def add_card(self, card: Card) -> None:
        self.cards.append(card)

    def print_hand(self) -> None:
        for card in self.cards:
            card.print_card()


if __name__ == "__main__":
    # Build a standard 52-card deck (faces 1..13 in each suit).
    all_cards: List[Card] = []
    for s in Suit:
        for face in range(1, 14):
            all_cards.append(SimpleCard(face, s))

    deck = Deck()
    deck.set_deck_of_cards(all_cards)
    deck.shuffle()  # faithful no-op shuffle
    print("Remaining cards:", deck.remaining_cards())

    hand = Hand()
    hand.add_card(SimpleCard(3, Suit.HEART))
    hand.add_card(SimpleCard(4, Suit.SPADE))
    print("Hand: ", end="")
    hand.print_hand()
    print()
    print("Score:", hand.score())

    dealt = deck.deal_hand(5)
    print("Dealt 5 cards: ", end="")
    if dealt:
        for c in dealt:
            if c is not None:
                c.print_card()
    print()

File: Coding/golang/deckofcards/deckofcards.go
package main

import (
	"fmt"
	"math/rand"
)

// Card models the abstract Java Card. value() is the abstract method that concrete
// cards must implement; the rest is shared behaviour provided by baseCard.
type Card interface {
	value() int
	getSuit() Suit
	isAvailable() bool
	markUnavailable()
	markAvailable()
	print()
}

// Suit enumerates the four card suits (value matches the Java enum ordinal).
type Suit int

const (
	Club Suit = iota
	Diamond
	Heart
	Spade
)

func (s Suit) getValue() int {
	return int(s)
}

// getSuitFromValue returns the Suit for a value, and false if the value is invalid.
func getSuitFromValue(value int) (Suit, bool) {
	switch value {
	case 0:
		return Club, true
	case 1:
		return Diamond, true
	case 2:
		return Heart, true
	case 3:
		return Spade, true
	default:
		return 0, false
	}
}

// baseCard holds the state and shared behaviour common to all cards.
type baseCard struct {
	available bool
	// faceValue is a number 2-10, 11 for Jack, 12 for Queen, 13 for King, or 1 for Ace.
	faceValue int
	suit      Suit
}

func newBaseCard(c int, s Suit) baseCard {
	return baseCard{available: true, faceValue: c, suit: s}
}

func (b *baseCard) getSuit() Suit       { return b.suit }
func (b *baseCard) isAvailable() bool    { return b.available }
func (b *baseCard) markUnavailable()     { b.available = false }
func (b *baseCard) markAvailable()       { b.available = true }

func (b *baseCard) print() {
	faceValues := []string{"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"}
	fmt.Print(faceValues[b.faceValue-1])
	switch b.suit {
	case Club:
		fmt.Print("c")
	case Heart:
		fmt.Print("h")
	case Diamond:
		fmt.Print("d")
	case Spade:
		fmt.Print("s")
	}
	fmt.Print(" ")
}

// simpleCard is a concrete Card used for demonstration. The reference Java Card is
// abstract with an abstract value(); this provides a minimal value() implementation.
type simpleCard struct {
	baseCard
}

func newSimpleCard(faceValue int, suit Suit) *simpleCard {
	return &simpleCard{newBaseCard(faceValue, suit)}
}

func (c *simpleCard) value() int { return c.faceValue }

// Deck is a deck of cards that deals from the top.
type Deck struct {
	cards      []Card
	dealtIndex int // marks first undealt card
}

func (d *Deck) setDeckOfCards(deckOfCards []Card) {
	d.cards = deckOfCards
}

func (d *Deck) shuffle() {
	for i := len(d.cards) - 1; i > 0; i-- {
		// Faithful port of Java `(int) Math.random() % (i + 1)`: truncating a value
		// in [0,1) to an int yields 0, so j is always 0 (reference bug preserved).
		j := int(rand.Float64()) % (i + 1)
		card1 := d.cards[i]
		card2 := d.cards[j]
		d.cards[i] = card2
		d.cards[j] = card1
	}
}

func (d *Deck) remainingCards() int {
	return len(d.cards) - d.dealtIndex
}

func (d *Deck) dealHand(number int) []Card {
	if d.remainingCards() < number {
		return nil
	}
	hand := make([]Card, number)
	count := 0
	for count < number {
		card := d.dealCard()
		if card != nil {
			hand[count] = card
			count++
		}
	}
	return hand
}

func (d *Deck) dealCard() Card {
	if d.remainingCards() == 0 {
		return nil
	}
	card := d.cards[d.dealtIndex]
	card.markUnavailable()
	d.dealtIndex++
	return card
}

func (d *Deck) print() {
	for _, card := range d.cards {
		card.print()
	}
}

// Hand is a collection of cards held by a player.
type Hand struct {
	cards []Card
}

func (h *Hand) score() int {
	score := 0
	for _, card := range h.cards {
		score += card.value()
	}
	return score
}

func (h *Hand) addCard(card Card) {
	h.cards = append(h.cards, card)
}

func (h *Hand) print() {
	for _, card := range h.cards {
		card.print()
	}
}

func main() {
	var cards []Card
	for v := 1; v <= 13; v++ {
		for s := 0; s < 4; s++ {
			suit, _ := getSuitFromValue(s)
			cards = append(cards, newSimpleCard(v, suit))
		}
	}

	deck := &Deck{}
	deck.setDeckOfCards(cards)
	deck.shuffle()

	fmt.Printf("Remaining cards: %d\n", deck.remainingCards())

	dealt := deck.dealHand(5)
	hand := &Hand{}
	for _, c := range dealt {
		hand.addCard(c)
	}

	fmt.Print("Dealt hand: ")
	hand.print()
	fmt.Println()
	fmt.Printf("Hand score: %d\n", hand.score())
}
